File: ADT/dfs_bfs/1971.py
# Microsoft, Amazon, Adobe

# Using DFS
# Tc = O(m + n)
from collections import defaultdict


class Solution:
    def check(self, mp, node, destination, visited):
        if node == destination:
            return True
        if visited[node]:
            return False
        visited[node] = True
        for it in mp[node]:
            if self.check(mp, it, destination, visited):
                return True
        return False

    def validPath(
        self, n: int, edges: List[List[int]], source: int, destination: int
    ) -> bool:
        if source == destination:
            return True
        mp = defaultdict(list)
        for u, v in edges:
            mp[u].append(v)
            mp[v].append(u)
        visited = [False] * n
        return self.check(mp, source, destination, visited)


# Using BFS
from collections import defaultdict, deque


class Solution:
    def validPath(
        self, n: int, edges: List[List[int]], source: int, destination: int
    ) -> bool:
        mp = defaultdict(list)
        for u, v in edges:
            mp[u].append(v)
            mp[v].append(u)
        visited = [False] * n
        que = deque([source])
        visited[source] = True
        while que:
            node = que.popleft()
            if node == destination:
                return True
            for neighbor in mp[node]:
                if not visited[neighbor]:
                    visited[neighbor] = True
                    que.append(neighbor)
        return False

File: ADT/dfs_bfs/1971.cpp
// Using DFS
class Solution {
public:
    bool check(unordered_map<int, vector<int>>& mp, int node, int destination,
               vector<bool>& visited) {
        if (node == destination)
            return true;
        if (visited[node])
            return false;
        visited[node] = true;
        for (auto& it : mp[node]) {
            if (check(mp, it, destination, visited))
                return true;
        }
        return false;
    }
    bool validPath(int n, vector<vector<int>>& edges, int source,
                   int destination) {
        if (source == destination)
            return true;
        unordered_map<int, vector<int>> mp;
        for (vector<int>& vec : edges) {
            int u = vec[0];
            int v = vec[1];
            mp[u].push_back(v);
            mp[v].push_back(u);
        }
        vector<bool> visited(n, false);
        return check(mp, source, destination, visited);
    }
};


// Using BFS
class Solution {
public:    
    bool validPath(int n, vector<vector<int>>& edges, int source, int destination) {
        unordered_map<int, vector<int>> mp;
        
        for(vector<int> &vec : edges) {
            int u = vec[0];
            int v = vec[1];
            
            mp[u].push_back(v);
            mp[v].push_back(u);
        }
        
        vector<bool> visited(n, false);
        
        queue<int> que;
        que.push(source);
        visited[source] = true;
        
        while(!que.empty()) {
            int node = que.front();
            que.pop();
            if(node == destination)
                return true;

            for(auto &it : mp[node]) {
                if(!visited[it]) {
                    visited[it] = true;
                    que.push(it);
                }
            }
        }
        
        return false;
    }
};